src/sqlfluff/dialects/dialect_mysql.py

```diff
@@ -129,21 +136,129 @@ class CreateTableStatementSegment(
     DelimiterSegment=OneOf(Ref("SemicolonSegment"), Ref("TildeSegment")),
     TildeSegment=StringParser(
         "~", SymbolSegment, name="tilde", type="statement_terminator"
     ),
+    ParameterNameSegment=RegexParser(
+        r"`?[A-Za-z0-9_]*`?", CodeSegment, name="parameter", type="parameter"
+    ),
+)
+
+mysql_dialect.patch_lexer_matchers(
+    [
+        RegexLexer(
+            "ampersand",
+            r"[@][a-zA-Z0-9_]*",
+            CodeSegment,
+        ),
+    ]
 )
 
 
+@mysql_dialect.segment()
+class DeclareStatement(BaseSegment):
+    """DECLARE statement.
+
+    mysql: https://dev.mysql.com/doc/refman/8.0/en/declare-local-variable.html
+    mysql: https://dev.mysql.com/doc/refman/8.0/en/declare-handler.html
+    mysql: https://dev.mysql.com/doc/refman/8.0/en/declare-condition.html
+    https://dev.mysql.com/doc/refman/8.0/en/declare-cursor.html
+    """
+
+    type = "declare_statement"
+
+    match_grammar = OneOf(
+        Sequence(
+            "DECLARE",
+            Ref("NakedIdentifierSegment"),
+            "CURSOR",
+            "FOR",
+            Ref("StatementSegment"),
+        ),
+        Sequence(
+            "DECLARE",
+            OneOf("CONTINUE", "EXIT", "UNDO"),
+            "HANDLER",
+            "FOR",
+            OneOf(
+                "SQLEXCEPTION",
+                "SQLWARNING",
+                Sequence("NOT", "FOUND"),
+                Sequence(
+                    "SQLSTATE",
+                    Ref.keyword("VALUE", optional=True),
+                    Ref("QuotedLiteralSegment"),
+                ),
+                OneOf(
+                    Ref("QuotedLiteralSegment"),
+                    Ref("NumericLiteralSegment"),
+                    Ref("NakedIdentifierSegment"),
+                ),
+            ),
+            Sequence(Ref("StatementSegment")),
+        ),
+        Sequence(
+            "DECLARE",
+            Ref("NakedIdentifierSegment"),
+            "CONDITION",
+            "FOR",
+            OneOf(Ref("QuotedLiteralSegment"), Ref("NumericLiteralSegment")),
+        ),
+        Sequence(
+            "DECLARE",
+            Ref("LocalVariableNameSegment"),
+            Ref("DatatypeSegment"),
+            Sequence(
+                Ref.keyword("DEFAULT"),
+                OneOf(
+                    Ref("QuotedLiteralSegment"),
+                    Ref("NumericLiteralSegment"),
+                    Ref("FunctionSegment"),
+                ),
+                optional=True,
+            ),
+        ),
+    )
+
+
+@mysql_dialect.segment(replace=True)
+class StatementSegment(ansi_dialect.get_segment("StatementSegment")):  # type: ignore
+    """Overriding StatementSegment to allow for additional segment parsing."""
+
+    parse_grammar = ansi_dialect.get_segment("StatementSegment").parse_grammar.copy(
+        insert=[
+            Ref("DelimiterStatement"),
+            Ref("CreateProcedureStatementSegment"),
+            Ref("TransactionStatementSegment"),
```

Comment:
Because you're replacing an existing segment (e.g. `TransactionStatementSegment`) you don't need to reinsert it here, it's already part of `StatementSegment`. You should only need to add segments which don't exist at all in the parent dialect.

Reply:
@alanmcruickshank  I got that cleaned up now